Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Payment.hs

```diff
@@ -217,34 +216,51 @@ notifyAndUpdateInvoiceStatusIfPaymentFailed driverId orderId orderStatus eventNa
   activeExecutionInvoice <- QIN.findByIdWithPaymenModeAndStatus (cast orderId) INV.AUTOPAY_INVOICE INV.ACTIVE_INVOICE
   let paymentMode = if isJust activeExecutionInvoice then DP.AUTOPAY else DP.MANUAL
   let (notifyFailure, updateFailure) = toNotifyFailure (isJust activeExecutionInvoice) eventName orderStatus
-  when (updateFailure || (not fromWebhook && notifyFailure)) $ do
-    QIN.updateInvoiceStatusByInvoiceId INV.FAILED (cast orderId)
-    case activeExecutionInvoice of
-      Just invoice' -> do
+  if updateFailure || (not fromWebhook && notifyFailure)
+    then do
+      QIN.updateInvoiceStatusByInvoiceId INV.FAILED (cast orderId)
+      whenJust activeExecutionInvoice $ \invoice' -> do
         QDF.updateAutoPayToManual invoice'.driverFeeId
         QDF.updateAutopayPaymentStageById (Just EXECUTION_FAILED) invoice'.driverFeeId
-      Nothing -> pure ()
-    nofityPaymentFailureIfNotNotified paymentMode
-
-  when (notifyFailure && isJust mbBankErrorCode) $
-    nofityPaymentFailureIfNotNotified paymentMode
+      nofityUltimatePaymentFailureIfNotNotifiedWithoutLimitAndError paymentMode
+    else when (notifyFailure && isJust mbBankErrorCode) $ do
+      nofityPaymentFailureIfNotNotified paymentMode
   where
     nofityPaymentFailureIfNotNotified paymentMode = do
-      let key = "driver-offer:FailedNotif-" <> orderId.getId
-      sendNotificationIfNotSent key 3600 $ fork "Sending payment failure notification" (PaymentNudge.notifyPaymentFailure driverId paymentMode mbBankErrorCode)
+      driver <- B.runInReplica $ QP.findById driverId >>= fromMaybeM (PersonDoesNotExist driverId.getId)
```

Comment:
you need not fetch driver for getting `merchantOperatingCityId`, you can pass `merchantOperatingCityId` from the parent functions.

Reply:
`notifyAndUpdateInvoiceStatusIfPaymentFailed` is called from two places. In one we have access to `merchantOperatingCityId` and from one we don't have `merchantOperatingCityId`. In the second case, we just have personId coming from juspayWebHook. So we need to fetch driver in order to get the `merchantOperatingCityId`